Fake Tensor Part 1

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78524
* #78523
* #78522
* #78517
* #78270
* #78090
* #78516
* #77972
* #77971
* #77970
* __->__ #77969

This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.

Commit:
Fake Tensor Part 1

[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

ezyang commented on test/test_fake_tensor.py:
```diff
@@ -0,0 +1,80 @@
+# Owner(s): ["module: unknown"]
```
well, could always chuck it in module: meta tensors

eellison replied:
`test_meta` is under `module: primTorch`.. I dont think there is a meta module

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -130,3 +132,11 @@ def _push_mode(ctor, mode_info: _ModeInfo) -> Iterator[object]:
+    try:
+        yield
+    finally:
+        del guard
```
THANKSSS

ezyang commented on torch/_subclasses/__init__.py:
```diff
@@ -0,0 +1,10 @@
+from torch._subclasses.base_tensor import BaseTensor
+from torch._subclasses.fake_tensor import FakeTensor, _device_not_kwarg_ops
+
+_all__ = [
```
missing leading underscore?

ezyang commented on torch/_subclasses/base_tensor.py:
```diff
@@ -0,0 +1,26 @@
+    # To ensure constructors can cooperate with one another, must accept and
+    # ignore element tensor (TODO: is this right???)
+    def __init__(self, elem):
+        super().__init__()
```
I'm pretty confident we can dump this constructor, it never actually got used for anything

ezyang commented on torch/_subclasses/base_tensor.py:
```diff
@@ -0,0 +1,26 @@
+    # typically must be disabled
+    __torch_function__ = torch._C._disabled_torch_function_impl
+
+__all__ = ["BaseTensor"]
```
My preference is to not add this class, and inline the only thing you need (in this case, ` __torch_function__ = torch._C._disabled_torch_function_impl`)

eellison replied:
My understanding from talking to various folks was the 
```
    @staticmethod
    def __new__(cls, elem, *, requires_grad=None):
        if requires_grad is None:
            return super().__new__(cls, elem)  # type: ignore
        else:
            return cls._make_subclass(cls, elem, requires_grad)
```
part was necessary..  I think I'll file issue for removal of this in future PR and leave for now

ezyang replied:
Huh, that's surprising.

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -0,0 +1,140 @@
+    aten.is_pinned.default,
+    aten.to.device,
+    aten.to.prim_Device,
+    aten._pin_memory.default,
```
@anjali411 Would this be a good operator tag? (Or maybe we should be able to figure this out from schema?)

anjali411 replied:
We can already infer this information from the FunctionSchema by iterating through the inputs
```
>>> torch.ops.aten.to.device._schema.arguments
[const Tensor& self, Device device, int dtype, bool non_blocking, bool copy, int? memory_format]
>>> torch.ops.aten.to.device._schema.arguments[1]
Device device
>>> torch.ops.aten.to.device._schema.arguments[1].kwarg_only
False
```

eellison replied:
Yeaaa I have a [test for that](https://github.com/pytorch/pytorch/blob/master/test/test_fake_tensor.py#L235). Maybe I just should just get rid of the list and move over the checking logic

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -0,0 +1,140 @@
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        assert elem.device.type == "meta"
+        device = device if isinstance(device, torch.device) else torch.device(device)
+        assert device.type != "meta"
```
You know... it might be ok for the inner device to be meta lol

eellison replied:
That seems like something has gone wrong (lost what the actual device is)... maybe if it proves to be needed in the future we can change

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -0,0 +1,140 @@
+    @staticmethod
+    def from_tensor(t):
+        existing_device = t.device
+        return FakeTensor(t.to(device="meta"), existing_device)
```
It probably doesn't matter for most users, but maybe it is necessary to do something fancier https://github.com/pytorch/pytorch/blob/50cadfae10fe61b6daa0d4f39a3a97fa19bc82eb/test/test_meta.py#L70-L203

eellison replied:
Resolved in follow up pr https://github.com/pytorch/pytorch/pull/78090

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618467).

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.

Differential Revision: [D36618467](https://our.internmc.facebook.com/intern/diff/D36618467)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618467).

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.

Differential Revision: [D36618467](https://our.internmc.facebook.com/intern/diff/D36618467)

[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.

Differential Revision: [D36618467](https://our.internmc.facebook.com/intern/diff/D36618467)

[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.

Differential Revision: [D36618467](https://our.internmc.facebook.com/intern/diff/D36618467)

[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.

Differential Revision: [D36618467](https://our.internmc.facebook.com/intern/diff/D36618467)

[ghstack-poisoned]

Commit:
Update on "Fake Tensor Part 1"


This is just copying over the PR from https://github.com/albanD/subclass_zoo/pull/32, and pulling in `BaseTensor`, plus adding one test based on invariants about schemas. 

From code comments
>  Meta tensors give you the ability to run PyTorch code without having to
 actually do computation through tensors allocated on a `meta` device.
 Because the device is `meta`, meta tensors do not model device propagation.
 FakeTensor extends MetaTensors to also carry an additional `fake_device`
 which tracks devices that would have been used.



[ghstack-poisoned]

eellison commented:
@pytorchbot merge this please
